### PR TITLE
Raise visibility of Out Of Memory errors, and avoid them for minified JavaScript files

### PR DESCRIPTION
**Pull request by maxjacobson**
* allow OOM errors to become engine errors (unreverts an earlier PR)
* skip minified JS files (algorithm copied mostly from https://github.com/codeclimate/codeclimate-eslint/pull/115) because our reproduction case from an OSS repo was a large minified JS file.
* print paths to stderr before analysis, so if these errors continue, we'll be able to see which file led to the error (makes test output messy at the moment, unfortunately)

@codeclimate/review

### CODE REVIEW

File: lib/cc/engine/analyzers/file_thread_pool.rb [hunk 1]
```diff
@@ -54,6 +54,13 @@ def thread_count
             DEFAULT_CONCURRENCY
           end
         end
+
+        def with_thread_abortion
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/file_thread_pool.rb]: Cringe... How about, `#with_thread_abort_on_exceptions`?

File: lib/cc/engine/analyzers/analyzer_base.rb [hunk 2]
```diff
@@ -20,7 +20,13 @@ def initialize(engine_config:)
         end
 
         def run(file)
-          process_file(file)
+          $stderr.puts("About to analyze #{file}. If analysis fails, consider excluding this file.")
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/analyzer_base.rb]: I would move this into the `else` so we don't show two messages in the skip case.

Comment (pbrisbin) [inline lib/cc/engine/analyzers/analyzer_base.rb]: I don't think we need to go so far as to say "If analysis fails", and this feels a little too wordy to me in general. Just saying `"Processing #{file}..."` is enough, IMO.

File: lib/cc/engine/analyzers/analyzer_base.rb [hunk 3]
```diff
@@ -20,7 +20,13 @@ def initialize(engine_config:)
...
+          skip, skip_reason = skip?(file)
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/analyzer_base.rb]: To avoid the multiple return values, I think I'd prefer this method return a reason or `nil`, so it's just

```rb
if (reason = skip?(file)`
```

File: lib/cc/engine/analyzers/javascript/minification_checker.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            line_count = 0
+            chars_seen = 0
+            content.chars.each do |char|
+              if char == newline
+                line_count += 1
+              else
+                chars_seen += 1
+              end
+            end
+            (chars_seen / line_count) >= MINIFIED_AVG_LINE_LENGTH_CUTOFF
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/javascript/minification_checker.rb]: 💄 can you separate the return value with a blank line?

**Comment by pbrisbin** [comment]
> makes test output messy at the moment, unfortunately

There's a bunch of ways we could quiet the specs, I think we should do that :)

**Comment by maxjacobson** [comment]
@pbrisbin good calls! Updated

**Comment by wfleming** [comment]
nit, but the PR title seems like the opposite of what we're doing: we're not avoiding the errors, we're increasing their visibility.

**Comment by maxjacobson** [comment]
@wfleming true enough. The message only really describes the second commit. I'll update

File: lib/cc/engine/analyzers/javascript/minification_checker.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        class MinificationChecker
+          MINIFIED_AVG_LINE_LENGTH_CUTOFF = 100
+
+          def initialize(path)
+            @content = File.read(path)
+          end
+
+          def minified?
+            line_count = content.lines.count
+            chars_seen = content.chars.count
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/javascript/minification_checker.rb]: How about just

```rb
ratio = content.chars.count / content.lines.count
ratio >= MINIFIED_AVG_LINE_LENGTH_CUTOFF
```

(Note: I rescind my "white space before return value" suggestion in the case of a 2-line method).

File: lib/cc/engine/analyzers/file_thread_pool.rb [hunk 6]
```diff
@@ -54,6 +54,13 @@ def thread_count
             DEFAULT_CONCURRENCY
           end
         end
+
+        def with_thread_abort_on_exceptions
+          t = Thread.new do
+            yield
+          end
+          (t.abort_on_exception = true) && t
```
Comment (pbrisbin) [inline lib/cc/engine/analyzers/file_thread_pool.rb]: Our style guide allows the use of `and` to avoid these parenthesis. That said, I'd prefer this formatting:

```rb
def with_thread_abort_on_exceptions(&block)
  thread = Thread.new(&block)
  thread.abort_on_exception = true
  thread
end
```

FWIW

**Comment by pbrisbin** [comment]
Some minor syntax suggestions, LGTM either way.

**Comment by maxjacobson** [comment]
Resolved conflicts and going to merge on green
